src/nvim/api/buffer.c

```diff
@@ -935,6 +969,17 @@ void nvim_buf_set_keymap(uint64_t channel_id, Buffer buffer, String mode, String
                          Dict(keymap) *opts, Error *err)
   FUNC_API_SINCE(6)
 {
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if (!buf) {
+    return;
+  }
+
+  // load buffer first if it's not loaded
+  if (!buf_ensure_loaded(buf)) {
+    api_set_error(err, kErrorTypeException, "Failed to load buffer");
+    return;
+  }
   modify_keymap(channel_id, buffer, false, mode, lhs, rhs, opts, err);
```

Comment:
I don't think neither keymap state nor variable (get/set_var) state depend on the buffer being loaded, so it can be omitted.